source/h1_connection.c

```diff
@@ -1066,6 +1080,27 @@ static int s_decoder_on_header(const struct aws_h1_decoded_header *header, void
                 connection->synced_data.new_stream_error_code = AWS_ERROR_HTTP_CONNECTION_CLOSED;
                 aws_h1_connection_unlock_synced_data(connection);
             } /* END CRITICAL SECTION */
+
+            /* For server side, server should finish sending the response. */
+            if (connection->base.client_data) {
+                /**
+                 * RFC-9112 section 9.6.
+                 * A client that receives a "close" connection option MUST cease sending
+                 *requests on that connection and close the connection after reading the
+                 *response message containing the "close" connection option.
```

Comment:
Ugh. Even in the new RFC this is ambiguous. Is it talking about "don't start any further requests" or saying "don't complete sending the current request"? You didn't capture the full sentence here. There's a semicolon and it goes on to talk specifically about pipelining, which makes me thing this means "don't start any further requests"

> A client that receives a "close" connection option MUST cease sending requests on that connection and close the connection after reading the response message containing the "close" connection option; if additional pipelined requests had been sent on the connection, the client SHOULD NOT assume that they will be processed by the server.

Reply:
Up in [section 9.5](https://www.rfc-editor.org/rfc/rfc9112#section-9.5-5) there's this:
> A client sending a message body SHOULD monitor the network connection for an error response while it is transmitting the request. If the client sees a response that indicates the server does not wish to receive the message body and is closing the connection, the client SHOULD immediately cease transmitting the body and close its side of the connection.

So ... we should definitely stop sending after a 404 Not Found "Connection: close"

BUT if we got a 100 Continue "Connection: close" 😳 we should NOT stop sending, because we still haven't received the full response.

But what about a 200 OK "Connection: close"? What's the right course of action there?

Reply:
Yeah, it's a bit ambiguous, but I think we can still just close connection and stop sending for those response. It will not hurt for the connection that server will stop sending anymore and only waits for the `corresponding close by the client`